src/PackagePrivate/DatabaseTermIdsResolver.php

```diff
@@ -85,31 +82,26 @@ private function selectTerms( IDatabase $db, array $termIds ): IResultWrapper {
 		);
 	}
 
-	private function preloadTypes( IResultWrapper $result ) {
+	private function loadTypes( IResultWrapper $result ) {
 		$typeIds = [];
 		foreach ( $result as $row ) {
-			$typeId = $row->wbtl_type_id;
-			if ( !array_key_exists( $typeId, $this->typeNames ) ) {
-				$typeIds[$typeId] = true;
-			}
+			$typeIds[$row->wbtl_type_id] = true;
 		}
-		$this->typeNames += $this->typeIdsResolver->resolveTypeIds( array_keys( $typeIds ) );
+		return $this->typeIdsResolver->resolveTypeIds( array_keys( $typeIds ) );
 	}
 
-	private function addResultTerms( array &$terms, stdClass $row ) {
-		$type = $this->lookupType( $row->wbtl_type_id );
+	private function addResultTerms( array &$terms, stdClass $row, array $types ) {
+		$typeId = $row->wbtl_type_id;
+		$type = $types[$typeId] ?? null;
+		if ( $type === null ) {
+			throw new InvalidArgumentException(
+				'Type ID ' . $typeId . ' was not found!' );
+		}
+
 		$lang = $row->wbxl_language;
 		$text = $row->wbx_text;
-		$terms[$type][$lang][] = $text;
-	}
 
-	private function lookupType( $typeId ) {
```

Comment:
Is there a reason why this is now inlined into `addResultTerms()`? I think it still makes sense to have it as a separate function (with an additional `$types` argument, if that’s where we’re going).

Reply:
the reason is that I wanted to throw the exception in the calling function in order to include the term id where it failed in the message.

I could've returned null in this one, and moved only the exception to the caller function, but then this function became simply a one line-wrapper so I thought I would as well remove it and move that line to the calling function as well since that calling function was itself also couple of lines of code.

I see your point about it being a separate function logically, but it didn't seem enough to justify one-liner function that is called only from one place (given that the exception is moved already out of it).

Reply:
Well in the current version the exception message only includes the type ID as far as I can see? (Though I might be missing something, GitHub’s UI for changes to pull request feels confusing sometimes.)